### DISCUSSION TITLE
Sign In via Functions

### DISCUSSION THREAD

**Discussion by TsFreddie** [supabase/supabase#14266] [Feature Requests]
This might be an unusual use case, video game platforms (steam, epic game store, google play, etc.) often provide their own session tickets which is really handy combining with Supabase Functions to verify users without manually providing credentials by the end users.

I'm struggling to find a way to create my own session and sign a jwt inside a function, so I'm guessing it might not be possible currently. 

Maybe something like `supabaseClient.auth.signInWithFunctions("steam", payload);` assisted with a function template that send request to the platform's API then sign and respond with a jwt would be helpful in these cases.

Having these kind of sign in methods available potientially open up Supabase to game developers to quickly develop a video game backend like inventory and maybe turn based matches.

**Comment by GaryAustin1** [comment]
You can generate a Supabase jwt on most server environments using a jwt library and the Supabase jwt secret key.  You can then set that jwt in the authorization header when you createClient and you will have access.   The jwt is only good until the expire time you set, so you would have to reset through some process on your side.

**Reply by ZelimDamian** [reply]
Have a similar use case where we need a custom auth flow and decided on functions. The issue with you suggestion is that `supabase.auth.setSession` wants both access/refresh tokens. But of course refresh token needs to be in `auth.refresh_tokens` table. We couldn't find any way to insert it in there through standard mechanisms (i.e. `supabase.select('refresh_tokens')` with schema set to 'auth' doesn't work)

**Reply by GaryAustin1** [reply]
@ZelimDamian, with the custom jwt approach you don't use setSession you just set the Authorization header when you createClient.

**Reply by ZelimDamian** [reply]
@GaryAustin1 Thanks! Will look into it. Will this disable the default refresh token handling and it would have to be handled manually?

**Reply by GaryAustin1** [reply]
This approach does not involve the session at all.  So if you are not regenerating the jwt on the fly, then yes you would need to deal with refreshing it if it needs it.  You should turn off persistSession and autoRefreshToken off (not sure they would do anything anyway without a session anyway).

**Comment by TsFreddie** [comment]
Thanks for the advice. I didn't end up testing signing my own jwts because I was having trouble getting users' uid from email or other custom columns via GoTrueAdminApi, and I don't know if the jwt will authenticate as the user with only email but no sub field.

But I did find out I can generate a magiclink and send the otp in the function and let the client use otp to sign in. I'm not exactly sure about the security implication but it seems to work.
